### DISCUSSION TITLE
Access Gate

### DISCUSSION THREAD

**Discussion by Marion5** [opcodesio/log-viewer#291] [General]
I want to authorize acces via Gate (https://log-viewer.opcodes.io/docs/3.x/configuration/access-to-log-viewer#via-viewlogviewer-gate)

In User I have a function isAdmin():
```
    public function isAdmin()
    {
        return true;
    }
```

I defined in my AuthServiceProvider:
```
        Gate::define('viewLogViewer', function (?User $user) {
            dump($user->isAdmin());
            return $user->isAdmin();
        });
```

1. The user has access to /log-viewer
2. The result of the dump is true
3. Beyound the headline "Log Viewer" is the following error: "Request failed with status code 500: Call to a member function isAdmin() on null" 

If I change my AuthServiceProvider:
```
        Gate::define('viewLogViewer', function (?User $user) {
            return true;
        });
```
it works.

Can you help me, why my solution with return $user-isAdmin(); does not work?

**Comment by arukompas** [comment]
hey @Marion5 ,

It looks likes the Log Viewer's API is not authenticated correctly. Here's a few things you might want to try:

1. Make sure the `config/log-viewer.php` contains the correct API middleware:
```php
    'api_middleware' => [
        \Opcodes\LogViewer\Http\Middleware\EnsureFrontendRequestsAreStateful::class,
        \Opcodes\LogViewer\Http\Middleware\AuthorizeLogViewer::class,
    ],
```

2. If you use any of the Laravel starter kits, such as Jetstream, Breeze, or use [Laravel Sanctum](https://laravel.com/docs/10.x/sanctum), please make sure you update your stateful domains ([see here](https://laravel.com/docs/10.x/sanctum#spa-configuration)). Keep in mind that `www.example.com` and `example.com` are separate domains in this case, so make sure to define the correct one, or use both, separated by comma.

Let me know how that works out!

**Comment by Marion5** [comment]
Hi @arukompas , 

i have checked the following:

config/log-viewer.php:

```
'route_domain' => null,
    'middleware' => [
        'web',
        \Opcodes\LogViewer\Http\Middleware\AuthorizeLogViewer::class,
    ],
    'api_middleware' => [
        \Opcodes\LogViewer\Http\Middleware\EnsureFrontendRequestsAreStateful::class,
        \Opcodes\LogViewer\Http\Middleware\AuthorizeLogViewer::class,
    ],
    'hosts' => [
        'local' => [
            'name' => ucfirst(env('APP_ENV', 'local')),
        ],
    ],
```

Kernel.php:

```
   protected $middlewareGroups = [
        'web' => [
            \App\Http\Middleware\EncryptCookies::class,
            \Illuminate\Cookie\Middleware\AddQueuedCookiesToResponse::class,
            \Illuminate\Session\Middleware\StartSession::class,
            // \Illuminate\Session\Middleware\AuthenticateSession::class,
            \Illuminate\View\Middleware\ShareErrorsFromSession::class,
            \App\Http\Middleware\VerifyCsrfToken::class,
            \Illuminate\Routing\Middleware\SubstituteBindings::class,
        ],

        'admin' => [
            \App\Http\Middleware\Admin::class
        ],

        'api' => [
            //\Laravel\Sanctum\Http\Middleware\EnsureFrontendRequestsAreStateful::class,
            'throttle:api',
            \Illuminate\Routing\Middleware\SubstituteBindings::class,
        ],
    ];

```

and I tested this for Kernel, without success: 


```
        'api' => [
            \Laravel\Sanctum\Http\Middleware\EnsureFrontendRequestsAreStateful::class,
            'throttle:api',
            \Illuminate\Routing\Middleware\SubstituteBindings::class,
        ],
```

**Reply by agramazio** [reply]
Same problem for me, using Sanctum.
Solved 

1. adding `'auth'` in section `api_middleware` in `config/log-viewer.php`:
```
    'api_middleware' => [
        'auth',
        \Opcodes\LogViewer\Http\Middleware\EnsureFrontendRequestsAreStateful::class,
        \Opcodes\LogViewer\Http\Middleware\AuthorizeLogViewer::class,
    ],
```
2. adding `SANCTUM_STATEFUL_DOMAINS` to `.env` file, like:
```
# Enable stateful domains
SANCTUM_STATEFUL_DOMAINS="192.168.38.20:8850"
```

Hope it helps!

**Reply by Marion5** [reply]
Thanks for your answert - adding SANCTUM_STATEFUL_DOMAINS to .env file solved my problem

**Reply by agus-wisnu** [reply]
what should the SANCTUM_STATEFUL_DOMAINS value?

**Reply by arukompas** [reply]
> what should the SANCTUM_STATEFUL_DOMAINS value?

The domain(s) you're running the app on.

This is only relevant when hosting the app somewhere. You'll likely serve it behind a domain or a subdomain, so this environment variable should contain those domains so it knows these are safe to pass around authentication cookies.

**Comment by arukompas** [comment]
Hey @Marion5 

1. Is this happening locally, or in production/staging?

2. What does your `config/sanctum.php` look like? And your `SANCTUM_STATEFUL` environment variable inside .env

Thanks!

**Reply by Marion5** [reply]
Sorry for my late answer, i did not have seen your answer.

It ist happening locally and on production

```
config/sanctum.php

use Laravel\Sanctum\Sanctum;
return [
    'stateful' => explode(',', env('SANCTUM_STATEFUL_DOMAINS', sprintf(
        '%s%s',
        'localhost,localhost:3000,127.0.0.1,127.0.0.1:8000,::1',
        Sanctum::currentApplicationUrlWithPort()
    ))),
    'guard' => ['web'],
    'expiration' => null,
    'middleware' => [
        'verify_csrf_token' => App\Http\Middleware\VerifyCsrfToken::class,
        'encrypt_cookies' => App\Http\Middleware\EncryptCookies::class,
    ],
];
```

There ist none SANCTUM_STATEFUL int the .env file

**Reply by arukompas** [reply]
hey @Marion5 , I'm sorry but I have no other ideas at the moment. Without being able to debug the app myself, it's very hard.

Do you have any way I can reproduce the issue myself?
